### PR TITLE
3821 term code system batch jobs

### PR DESCRIPTION
**Pull request by TipzCM**
* Migrated Term Code System Delete and Term Code System Delete Version jobs to Batch2
* added changelog

### CODE REVIEW

**Review by jdar8 — COMMENTED** [review]
Looks good to me!

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/Batch2JobHelper.java [hunk 1]
```diff
@@ -118,9 +123,14 @@ public long getCombinedRecordsProcessed(String theJobId) {
...
-	// TODO KHS I don't think this works yet
-	public void awaitAllCompletions(String theJobDefinitionId) {
-		List<JobInstance> instances = myJobCoordinator.getInstancesbyJobDefinitionIdAndEndedStatus(theJobDefinitionId, false, 100, 0);
+	public void awaitAllJobsOfJobIdToComplete(String theJobDefinitionId) {
+		// fetch all jobs of any status type
+		List<JobInstance> instances = myJobCoordinator.getJobInstancesByJobDefinitionIdAndStatuses(
+			theJobDefinitionId,
+			new HashSet<>(Arrays.asList(StatusEnum.values())),
+			100,
+			0);
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/Batch2JobHelper.java:132]: will do for batchsize, but 0 is the page start and the default value of everything, so i'll suggest we leave that one.

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/Batch2JobHelper.java [hunk 2]
```diff
@@ -118,9 +125,14 @@ public long getCombinedRecordsProcessed(String theJobId) {
...
 	}
 
-	// TODO KHS I don't think this works yet
-	public void awaitAllCompletions(String theJobDefinitionId) {
-		List<JobInstance> instances = myJobCoordinator.getInstancesbyJobDefinitionIdAndEndedStatus(theJobDefinitionId, false, 100, 0);
+	public void awaitAllJobsOfJobIdToComplete(String theJobDefinitionId) {
+		// fetch all jobs of any status type
+		List<JobInstance> instances = myJobCoordinator.getJobInstancesByJobDefinitionIdAndStatuses(
+			theJobDefinitionId,
+			new HashSet<>(Arrays.asList(StatusEnum.values())),
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/Batch2JobHelper.java:132]: I really don't like this pattern.  Much prefer to just add a new method `getJobInstancesByJobDefinitionId` that omits status from the query criteria

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/term/job/TermCodeSystemVersionDeleteJobTest.java [hunk 3]
```diff
@@ -128,15 +123,21 @@ public void runDeleteJobDeleteOneVersion() throws Exception {
 		});
 
 
-		JobParameters jobParameters = new JobParameters(Collections.singletonMap(
-			JOB_PARAM_CODE_SYSTEM_VERSION_ID, new JobParameter(termCodeSystemVersionPidVect[0], true) ));
+//		JobParameters jobParameters = new JobParameters(Collections.singletonMap(
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/term/job/TermCodeSystemVersionDeleteJobTest.java:126]: commented code

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/TermCodeSystemJobConfig.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+			.errorHandler(deleteCodeSystemVersionCompletionHandler())
+			.build();
+	}
+
+	/**
+	 * Delete Code System Job
+	 * Deletes all code system versions, before deleting the code system itself
+	 */
+	@Bean
+	public JobDefinition<TermCodeSystemDeleteJobParameters> termCodeSystemDeleteJobDefinition() {
```
Comment (fil512) [inline hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/TermCodeSystemJobConfig.java:62]: nice keeping both of these in the same context

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/TermCodeSystemJobConfig.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	/** Delete code system version job **/
+	
+	@Bean
+	public DeleteCodeSystemVersionParameterValidator deleteCodeSystemVersionPrameterValidator() {
+		return new DeleteCodeSystemVersionParameterValidator();
+	}
+
+	@Bean
+	public DeleteCodeSystemVersionFirstStep deleteCodeSystemVersionFirstStep() {
+		return new DeleteCodeSystemVersionFirstStep();
```
Comment (fil512) [inline hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/TermCodeSystemJobConfig.java:141]: I'm worried with all these null constructors that means some of these beans contain @Autowire.  I've been trying to replace all @Autowired fields with final constructor fields as I come across them.  But I get if we just need to get this merged...

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/codesystemdelete/DeleteCodeSystemCompletionHandler.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import ca.uhn.fhir.batch2.api.IJobCompletionHandler;
+import ca.uhn.fhir.batch2.api.JobCompletionDetails;
+import ca.uhn.fhir.jpa.term.api.ITermCodeSystemDeleteJobSvc;
+import ca.uhn.fhir.jpa.term.models.TermCodeSystemDeleteJobParameters;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class DeleteCodeSystemCompletionHandler
+	implements IJobCompletionHandler<TermCodeSystemDeleteJobParameters> {
+
+	@Autowired
```
Comment (fil512) [inline hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/codesystemdelete/DeleteCodeSystemCompletionHandler.java:12]: recommend replacing autowired field with final field set by constructor

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/codesystemdelete/DeleteCodeSystemStep.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+		theDataSink.accept(new VoidModel()); // nothing required here
+
+		return RunOutcome.SUCCESS;
+	}
+
+	@NotNull
+	@Override
+	public ChunkOutcome consume(ChunkExecutionDetails<TermCodeSystemDeleteJobParameters, CodeSystemVersionPIDResult> theChunkDetails) {
+		/*
+		 * A single code system can have multiple versions.
```
Comment (fil512) [inline hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/codesystemdelete/DeleteCodeSystemStep.java:41]: interesting...

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/codesystemversiondelete/DeleteCodeSystemVersionCompletionHandler.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class DeleteCodeSystemVersionCompletionHandler implements IJobCompletionHandler<TermCodeSystemDeleteVersionJobParameters> {
+
+	@Autowired
+	private ITermCodeSystemDeleteJobSvc myTermCodeSystemSvc;
+
+	@Override
+	public void jobComplete(JobCompletionDetails<TermCodeSystemDeleteVersionJobParameters> theDetails) {
+		myTermCodeSystemSvc.notifyJobComplete(theDetails.getInstance().getInstanceId());
```
Comment (fil512) [inline hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/codesystemversiondelete/DeleteCodeSystemVersionCompletionHandler.java:16]: what does this do?

Comment (fil512) [reply hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/codesystemversiondelete/DeleteCodeSystemVersionCompletionHandler.java:16]: oh it's to clean up that redundant map i bet.  i really want to yank all this stuff if there's no strong reason for it

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/term/models/CodeSystemConceptsDeleteResult.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
+package ca.uhn.fhir.jpa.term.models;
+
+public class CodeSystemConceptsDeleteResult {
```
Comment (fil512) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/term/models/CodeSystemConceptsDeleteResult.java:3]: feels like this should be an IModelJson and the return value of the reducer (and stored in the job instance as job output effectively)

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/api/TermCodeSystemDeleteJobSvc.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.text.DecimalFormat;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+@Transactional()
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/api/TermCodeSystemDeleteJobSvc.java:23]: done

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/Batch2JobHelper.java [hunk 11]
```diff
@@ -118,9 +125,14 @@ public long getCombinedRecordsProcessed(String theJobId) {
 
 	}
 
-	// TODO KHS I don't think this works yet
-	public void awaitAllCompletions(String theJobDefinitionId) {
-		List<JobInstance> instances = myJobCoordinator.getInstancesbyJobDefinitionIdAndEndedStatus(theJobDefinitionId, false, 100, 0);
+	public void awaitAllJobsOfJobIdToComplete(String theJobDefinitionId) {
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/Batch2JobHelper.java:128]: changed

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/Batch2JobHelper.java (hunk 2 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/Batch2JobHelper.java:132]: chngd

File: hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/Batch2JobHelper.java [hunk 12]
```diff
@@ -52,7 +57,9 @@ public JobInstance awaitJobCompletion(Batch2JobStartResponse theStartResponse) {
 	}
 
 	public JobInstance awaitJobCompletion(String theId) {
-		await().until(() -> {
+		await()
+			.atMost(1, TimeUnit.MINUTES)
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-jpaserver-test-utilities/src/main/java/ca/uhn/fhir/jpa/test/Batch2JobHelper.java:61]: Ah, test code. It was always timing out so i had to stick this in temporarily.

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/term/job/TermCodeSystemVersionDeleteJobTest.java (hunk 3 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/term/job/TermCodeSystemVersionDeleteJobTest.java:126]: **remoed**

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/TermCodeSystemJobConfig.java (hunk 5 above)
Comment (TipzCM) [reply hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/TermCodeSystemJobConfig.java:141]: updated

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/codesystemversiondelete/DeleteCodeSystemVersionCompletionHandler.java (hunk 8 above)
Comment (TipzCM) [reply hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/codesystemversiondelete/DeleteCodeSystemVersionCompletionHandler.java:16]: let's discuss this outside of this review

File: hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/codesystemdelete/DeleteCodeSystemCompletionHandler.java (hunk 6 above)
Comment (TipzCM) [reply hapi-fhir-storage-batch2-jobs/src/main/java/ca/uhn/fhir/batch2/jobs/termcodesystem/codesystemdelete/DeleteCodeSystemCompletionHandler.java:12]: done

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/term/models/CodeSystemConceptsDeleteResult.java (hunk 9 above)
Comment (TipzCM) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/term/models/CodeSystemConceptsDeleteResult.java:3]: Is it needed though?

I kept the values cause the old system passed them back (and i think it's 'good practice' to return data that is available and let the consumer decide if they need it). But in this case, the delete result is not used (now?)

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermDeferredStorageSvcImpl.java [hunk 13]
```diff
@@ -68,22 +66,24 @@
...
+	private final List<TermConcept> myDeferredConcepts = Collections.synchronizedList(new ArrayList<>());
+	private final List<ValueSet> myDeferredValueSets = Collections.synchronizedList(new ArrayList<>());
+	private final List<ConceptMap> myDeferredConceptMaps = Collections.synchronizedList(new ArrayList<>());
+	private final List<TermConceptParentChildLink> myConceptLinksToSaveLater = Collections.synchronizedList(new ArrayList<>());
 
+	/**
+	 * A list of job ids for CodeSydstemDelete and CodeSystemVersionDelete jobs that
+	 * have been scheduled (but not completed)
+	 */
+	private final List<String> myJobExecutions = Collections.synchronizedList(new ArrayList<>());
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermDeferredStorageSvcImpl.java:86]: as discussed, added a todo

**Pull request merged**